src/Toml/Parser/Value.hs

```diff
@@ -148,22 +146,20 @@ tableNameP = between (text "[") (text "]") keyP
 -- Values
 
 decimalP :: Parser Integer
-decimalP = mkInteger <$> decimalStringP
+decimalP = zero <|> more
   where
-    decimalStringP   = fst <$> match (some digitChar >> many _digitsP)
-    _digitsP         = try (char '_') >> some digitChar
-    mkInteger        = textToInt . stripUnderscores
-    textToInt        = fst . fromRight (error "Underscore parser has a bug") . TR.decimal
-    stripUnderscores = Text.filter (/= '_')
+    zero = 0 <$ char '0'
+    more = check =<< readMaybe . concat <$> sepBy1 (some digitChar) (char '_')
+    check = maybe (fail "Not an integer") return
```

Comment:
@jiegillet Could you benchmark this PR before your PR and after your PR and paste the results here? I wonder what is the difference in performance.

Reply:
I benchmarked a toml file consisting only of `ints = [0, 1, 2, 120, 9000, -5, +12, -100, 500, +876]
`.
Current implementation:
```
benchmarked Parse/tomland
time                 53.71 μs   (43.82 μs .. 61.27 μs)
                     0.884 R²   (0.835 R² .. 0.949 R²)
mean                 41.11 μs   (39.23 μs .. 43.75 μs)
std dev              7.469 μs   (4.910 μs .. 10.92 μs)
variance introduced by outliers: 84% (severely inflated)
```

After change 

```
benchmarked Parse/tomland
time                 54.02 μs   (50.34 μs .. 59.31 μs)
                     0.969 R²   (0.943 R² .. 0.998 R²)
mean                 53.35 μs   (52.56 μs .. 54.87 μs)
std dev              4.032 μs   (2.875 μs .. 5.893 μs)
variance introduced by outliers: 50% (moderately inflated)
```
Doesn't look very different

Reply:
Thanks for teaching me how to benchmark @willbasky

Reply:
Although now I'm thinking that changing a parser potentially changes the behavior of the whole thing, so here is another benchmark of the original toml file + the `ints = [...]` from above.
Before change:
```
benchmarked Parse/tomland
time                 287.0 μs   (274.1 μs .. 299.8 μs)
                     0.986 R²   (0.978 R² .. 0.992 R²)
mean                 299.9 μs   (293.9 μs .. 307.5 μs)
std dev              21.87 μs   (17.97 μs .. 27.69 μs)
variance introduced by outliers: 46% (moderately inflated)
```

After change:
```
benchmarked Parse/tomland
time                 364.9 μs   (338.1 μs .. 390.5 μs)
                     0.973 R²   (0.961 R² .. 0.983 R²)
mean                 319.4 μs   (311.8 μs .. 331.4 μs)
std dev              31.78 μs   (26.91 μs .. 38.29 μs)
variance introduced by outliers: 60% (severely inflated)
```
That's a bigger change...

Reply:
Oh, that's quite a lot I guess. Seems like this feature is costly. How do you guys think?

Reply:
Yeah, I personally think that it's not worth it to forbid leading zeroes for such performance cost. It's like, not an addition of some new feature, it's more work and complication of parser for not so great benefit. I can understand why TOML specification forbids leading zeroes. But I'm not sure that we should support every bit of TOML spec. Performance is probably not a problem, it's still fast and TOML parsing is supposed to be run only once at the program start. I'm mostly worried about code obfuscation. I already don't understand some places of parser code :disappointed:

Reply:
(@ChShersh you will cry when you see the parser for arrays of tables then :p )
It's a 6% decrease in speed. I'm quite surprise it's that much. Maybe I can play around more, it's my first time using benchmarks.
I thought my implementation was less obfuscated than the previous one, to be honest (if you forget about `zero`, it's 2 lines instead of 5).

Reply:
@jiegillet I guess we can live with 6% performance decrease for array of tables :slightly_smiling_face: This is a very important feature. But 20% for not allowing leading zeroes... Way to much! :slightly_smiling_face: And I think that parser for array of tables should be manageable, this part we can handle :+1:

Reply:
What if, instead of having a 6% increase, I had a 4% _decrease_ with zero more complexity to the code, would you merge the feature then? :)

Reply:
@jiegillet This looks proimising! :smirk: But be careful in your benchmarks. Sometimes you see that the results _were inflated_. This means that the numbers are not precise and accurate enough.